paper-server/src/main/java/io/papermc/paper/adventure/providers/ClickCallbackProviderImpl.java

```diff
@@ -1,29 +1,38 @@
 package io.papermc.paper.adventure.providers;
 
+import io.papermc.paper.adventure.PaperAdventure;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.Queue;
 import java.util.UUID;
+import java.util.concurrent.ConcurrentLinkedQueue;
 import net.kyori.adventure.audience.Audience;
+import net.kyori.adventure.key.Key;
 import net.kyori.adventure.text.event.ClickCallback;
 import net.kyori.adventure.text.event.ClickEvent;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.Tag;
+import net.minecraft.resources.ResourceLocation;
 import org.jetbrains.annotations.NotNull;
 
-import java.util.Queue;
-import java.util.concurrent.ConcurrentLinkedQueue;
-
 @SuppressWarnings("UnstableApiUsage") // permitted provider
 public class ClickCallbackProviderImpl implements ClickCallback.Provider {
+    private static final Key CLICK_CALLBACK_KEY = Key.key("paper", "click_callback");
+    private static final String ID_KEY = "id";
 
+    public static final ResourceLocation CLICK_CALLBACK_RESOURCE_LOCATION = PaperAdventure.asVanilla(CLICK_CALLBACK_KEY);
     public static final CallbackManager CALLBACK_MANAGER = new CallbackManager();
 
     @Override
     public @NotNull ClickEvent create(final @NotNull ClickCallback<Audience> callback, final ClickCallback.@NotNull Options options) {
-        return ClickEvent.runCommand("/paper:callback " + CALLBACK_MANAGER.addCallback(callback, options));
+        final CompoundTag tag = new CompoundTag();
+        tag.putString(ID_KEY, CALLBACK_MANAGER.addCallback(callback, options).toString());
+        return ClickEvent.custom(CLICK_CALLBACK_KEY, PaperAdventure.asBinaryTagHolder(tag));
     }
 
     public static final class CallbackManager {
 
-        private final Map<UUID, StoredCallback> callbacks = new HashMap<>();
+        private final Map<UUID , StoredCallback> callbacks = new HashMap<>();
```

Comment:
wow extra space I can't believe it
```suggestion
        private final Map<UUID, StoredCallback> callbacks = new HashMap<>();
```